### PR TITLE
pkg/uprobetracer: add support for USDT tracepoints

### PR DESCRIPTION
**Pull request by i-Pear**
# pkg/uprobetracer: add support for USDT tracepoints

Added USDT support for uprobe ebpf programs.
A parser for USDT elf notes is included.
Docs are updated.

## How to use

Use `usdt/<programPath or libraryName>:<providerName>:<probeName>` as the section name of ebpf file.

## Testing

Use `SEC("usdt//usr/bin/python3:python:function__entry")` and `SEC("usdt//usr/bin/python3:python:function__return")`.

Then run a python program inside the container:

```
from time import sleep

def f():
    print(123)

while True:
    f()
    sleep(1)
```

### CODE REVIEW

File: pkg/uprobetracer/usdt.go
Comment (alban) [inline pkg/uprobetracer/usdt.go:110]:
```suggestion
			return nil, fmt.Errorf("reading USDT note header: %w", err)
```

ditto below: reading, seeking...

**Review by alban — COMMENTED** [review]
Thanks. Some initial comments.

Could you add a gadget for python `function__entry` / `function__return`?

**Comment by i-Pear** [comment]
> Could you add a gadget for python function__entry / function__return?

I'm not sure if this is practical. Since we currently cannot elegantly read the parameters, such a gadget can only output "func is called" and "func returned", but it cannot indicate which function it is.

**Comment by alban** [comment]
> > Could you add a gadget for python function__entry / function__return?
> 
> I'm not sure if this is practical. Since we currently cannot elegantly read the parameters, such a gadget can only output "func is called" and "func returned", but it cannot indicate which function it is.

Isn't it the same as with uprobes?

The uprobe function receives a `struct pt_regs *ctx` as argument and from there it can read the arguments of the tracepoint. If a pointer needs to be dereferenced, the gadget can use bpf helper functions like `bpf_probe_read_user()`.

But I am not sure how it chooses the order of the registers:
```
    DTRACE_PROBE3(tracetest, testprobe2, "foo", "bar", "bar3");
```

```
  stapsdt              0x00000042	NT_STAPSDT (SystemTap probe descriptors)
    Provider: tracetest
    Name: testprobe2
    Location: 0x0000000000401155, Base: 0x000000000040201d, Semaphore: 0x0000000000000000
    Arguments: 8@%rax 8@%rdx 8@%rcx
```

The python gadget and parameter support can be done in a later PR.

**Comment by alban** [comment]
The usdt program is using a `struct pt_regs*` as parameter:
```
struct pt_regs;
#define PT_REGS_PARM1(x)...
#define PT_REGS_PARM2(x)...
```
We could define our own `struct gadget_usdt_parms` and relocate the fields to match the fields in `struct pt_regs` with the order specified in the `.note.stapsdt` section (e.g. "8@%rax 8@%rdx 8@%rcx").

But in practice, software using /usr/include/sys/sdt.h are using `#define _SDT_ARGTMPL(id)       %[id]` for which the compiler automatically generate "%rax" for argument 1, "%rdx" for argument 2, etc.

So if the gadget hard codes `PT_REGS_PARM1(x)`, `PT_REGS_PARM2(x)`, etc. that would work in practice for software using `DTRACE_PROBE1`, `DTRACE_PROBE2`...

**Comment by i-Pear** [comment]
I'm using `<sys/sdt.h>, but it seems not having such behaviour:

In my `test.cpp`:
```
        int a,b,c,d,e,f,g;
        STAP_PROBE(prov, prob);
        STAP_PROBE1(prov, prob, a);
        STAP_PROBE2(prov, prob, a, b);
        STAP_PROBE3(prov, prob, a, b, c);
        STAP_PROBE4(prov, prob, a, b, c, d);
        STAP_PROBE5(prov, prob, a, b, c, d, e);
        STAP_PROBE6(prov, prob, a, b, c, d, e, f);
```
I use `g++` to compile, and the `readelf -n` says:
```
Displaying notes found in: .note.stapsdt
  Owner                Data size 	Description
  stapsdt              0x00000023	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001131, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: 
  stapsdt              0x0000002f	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001132, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@-24(%rbp)
  stapsdt              0x0000003c	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001133, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@-24(%rbp) -4@-20(%rbp)
  stapsdt              0x00000049	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001134, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@-24(%rbp) -4@-20(%rbp) -4@-16(%rbp)
  stapsdt              0x00000056	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001135, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@-24(%rbp) -4@-20(%rbp) -4@-16(%rbp) -4@-12(%rbp)
  stapsdt              0x00000062	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001136, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@-24(%rbp) -4@-20(%rbp) -4@-16(%rbp) -4@-12(%rbp) -4@-8(%rbp)
  stapsdt              0x0000006e	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001137, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@-24(%rbp) -4@-20(%rbp) -4@-16(%rbp) -4@-12(%rbp) -4@-8(%rbp) -4@-4(%rbp)
```

When I add `O2` option to `g++`, the `readelf` output turns to:
```
Displaying notes found in: .note.stapsdt
  Owner                Data size 	Description
  stapsdt              0x00000023	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001044, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: 
  stapsdt              0x0000002a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001047, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax
  stapsdt              0x00000032	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x000000000000104a, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx
  stapsdt              0x0000003a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x000000000000104d, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx
  stapsdt              0x00000042	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001050, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx -4@%esi
  stapsdt              0x0000004a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001053, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx -4@%esi -4@%edi
  stapsdt              0x00000050	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001054, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx -4@%esi -4@%edi -4@$0
```

> But in practice, software using /usr/include/sys/sdt.h are using #define _SDT_ARGTMPL(id)       %[id] for which the compiler automatically generate "%rax" for argument 1, "%rdx" for argument 2, etc.

Are these results all with `-O2`? I'm not sure what it's behaviour in more complex programs...

**Comment by i-Pear** [comment]
An extra simple test, I reversed the arguments in `PROBE6`:
```
        int a,b,c,d,e,f,g;
        STAP_PROBE(prov, prob);
        STAP_PROBE1(prov, prob, a);
        STAP_PROBE2(prov, prob, a, b);

        STAP_PROBE3(prov, prob, a, b, c);
        STAP_PROBE4(prov, prob, a, b, c, d);
        STAP_PROBE5(prov, prob, a, b, c, d, e);
        STAP_PROBE6(prov, prob, f, e, d, c, b, a);
```
And the `readelf` says:
```
Displaying notes found in: .note.stapsdt
  Owner                Data size 	Description
  stapsdt              0x00000023	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001044, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: 
  stapsdt              0x0000002a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001047, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax
  stapsdt              0x00000032	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x000000000000104a, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx
  stapsdt              0x0000003a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x000000000000104d, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx
  stapsdt              0x00000042	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001050, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx -4@%esi
  stapsdt              0x0000004a	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001053, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@%eax -4@%edx -4@%ecx -4@%esi -4@%edi
  stapsdt              0x00000050	NT_STAPSDT (SystemTap probe descriptors)
    Provider: prov
    Name: prob
    Location: 0x0000000000001054, Base: 0x0000000000002004, Semaphore: 0x0000000000000000
    Arguments: -4@$0 -4@%edi -4@%esi -4@%ecx -4@%edx -4@%eax
```
You can see the arguments are also reversed, so the patten `"%rax" for argument 1, "%rdx" for argument 2` is no longer working:(

**Comment by alban** [comment]
I was testing without `-O` options.
Thanks for checking. It is more complicated than I thought.

File: pkg/uprobetracer/usdt.go
Comment (alban) [inline pkg/uprobetracer/usdt.go:87]: Check if the section has the correct type:
```golang
if noteSection.Type != elf.SHT_NOTE {
    return nil, fmt.Errorf("section %q is not a note", sdtNoteSectionName)
}
```

Ditto below with `baseSection.Type != elf.SHT_PROGBITS`.

File: pkg/uprobetracer/usdt.go
Comment (alban) [inline pkg/uprobetracer/usdt.go:112]: You should read the name instead of skipping it, and check that it is the vendor string "stapsdt" and that `header.Type == 3` (for "sdt v3").

And if that's not the case, ignore and continue to the next entry in the for loop.

See https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation#SDT_Notes

File: pkg/uprobetracer/usdt.go
Comment (alban) [inline pkg/uprobetracer/usdt.go:49]: It does not seem to be an internal error in ig but a malformed elf file?

File: pkg/uprobetracer/usdt.go
Comment (alban) [inline pkg/uprobetracer/usdt.go:141]: The function `readStringFromBytes` could be moved from `ldcache_parser.go` to a new file `pkg/uprobetracer/bytes.go` since it is used both by `ldcache_parser.go` and `usdt.go`, and it is not related to ldcache.

**Review by alban — COMMENTED** [review]
Some small changes to do, it should be ready to go after that.

File: pkg/uprobetracer/tracer.go
Comment (i-Pear) [reply pkg/uprobetracer/tracer.go:254]: But using `Errorf` will change the log level to `error`? Failure while attaching to uprobes are common, because the target files may not have the required symbols, and in such cases we just ignore them. If the user want to know why some files are not attached, he can enable the verbose mode.

Comment (alban) [reply pkg/uprobetracer/tracer.go:254]: Ok, please keep Debugf then :)

Even with Debugf, `%q..., err.Error()` should be changed to `%w..., err`.

Comment (i-Pear) [reply pkg/uprobetracer/tracer.go:254]: Seems not, my Goland IDE says "This verb can be used only in 'fmt.Errorf' calls (%w) ". 
And the output with `%w` is:
DEBU[0000] failed to attach uprobe "trace_uprobe_free": %!w(*errors.errorString=&{123})

Comment (alban) [reply pkg/uprobetracer/tracer.go:254]: Oops, sorry for the wrong comment.

File: pkg/uprobetracer/bytes.go
Comment (alban) [inline pkg/uprobetracer/bytes.go:1]: +license header

**Review by alban — APPROVED** [review]
LGTM, thanks!

I'll wait for the CI to complete before merging.

TODO for future PRs:
- [ ] Handling USDT parameters
- [ ] Make searchForLibrary safe with TOCTOU symlink swap issues
- [ ] Adding some gadgets using USDT

**Pull request merged**
